back/app/models/basket.rb

```diff
@@ -46,6 +46,15 @@ def total_votes
     baskets_ideas.pluck(:votes).sum
   end
 
+  def destroy_or_keep!
+    if submitted_at.present? && participation_context_type == 'Phase' && participation_context.end_at <= Time.now
+      update!(user: nil)
+    else
+      destroy!
+      SideFxBasketService.new.after_destroy self, nil
```

Comment:
I think the overall approach is good, but my feeling is that we might want to move this logic to the application layer. For example from the `after_destroy` of `SideFxUserService`. This would keep the models simple and avoid dependencies from the models to a service.

The disadvantage is that we no longer have a strong guarantee that the baskets are deleted when users are deleted. I think this is acceptable if the logic happens in `after_destroy` (we should be allowed to assume that this will always be called on the right moment). And if we really want a stronger guarantee, we could add a model validation that we cannot have a basket without a user if the phase is not finished (although this would make phases invalid if an admin turns a past phase into an active phase, but then we should probably prohibit this).

Reply:
Today, we're already assuming that in the `after_destroy` of SideFxUserService is always called when deleting a user. For example, see the `DeleteUserJob`. Here we assume this to guarantee the user's privacies, which is also quite a big deal.

Reply:
Just to check - are you saying move all the logic for destroying baskets to the SideFxUserService? Or just the counts? Feels like it should be all of it.

Reply:
Or everything in the user model anyway.

Reply:
Yes all of it (deleting baskets and fixing the counts), but the `SideFxUserService` could delegate this to some other service, for example.